### ISSUE TITLE
Subscribe promise always succeeds after automatic reconnection to Faye server

### ISSUE THREAD

**Issue by robinroestenburg** [faye/faye#381]
Consider the following scenario I am trying to get working in my project:

1. Browser client connects to Faye server
1. Browser client subscribes to channel `/foo` using some kind of server-generated signature `x` (that is added to the `/meta/subscribe` message using an extension). The signature expires after X seconds (where X is lower than the server timeout).
1. Faye server goes down for a period that is longer than the timeout, causing the client to be disconnected on the server.
1. Faye server is back up.
1. Browser client automatically attempts to reconnect and succeeds.
1. Browser client automatically attempts to re-subscribe to `/foo` using signature `x`.
1. Faye server rejects the subscription attempt as the signature has expired.
1. This triggers the client to retrieve a new signature from the server and try again.
 
In step 6, the subscription is performed using:

``` js
client.subscribe('/foo', function(message) {
  // process message
}).then(
  function() { console.log('success'); },
  function() {
    console.log('failed');
    // renew signature and retry
  });
```
Using the returned promise I am able to detect the reject of the server and perform steps 7 and 8.

Unfortunately, this does not work as I expected:
* At step 5 the Faye client will retry to subscribe to the previously subscribed channels: https://github.com/faye/faye/blob/master/javascript/protocol/client.js#L103.
* After that, step 6 will be performed. The promise returned by `Faye.Client.subscribe()` will automatically be resolved successfully because of this code: https://github.com/faye/faye/blob/master/javascript/protocol/client.js#L205. The subscriptions will be performed in the background, but fail and giving me no way to retry them. 

I've added the following hack in my code, to work around this problem:

``` js
this.fayeClient.on('transport:up', function() {
  // Activate all registered subscriptions when the connection with
  // websockets server has been established.
  Object.keys(self.subscriptions).forEach(function(key) {
    self.subscription(key).activate();
  });
});
this.fayeClient.on('transport:down', function() {
  self.fayeClient._channels = new Faye.Channel.Set();
});
```

This way, `Faye.Client.handshake` will not re-subscribe to the previously registered channels, but I now have to do this manually (in `transport:up` event).

This works, but I would prefer a way to configure this in the Faye client.

Would it be an option to have the subscribe promise to be resolved by the response from the Faye server?

**Comment by jcoglan** [comment]
I'm not sure why step 6 is necessary, given the Faye client automatically reinstates its subscriptions for you. Using `transport:{up,down}` for this won't work in general, because those events don't necessarily correspond to a client having its session terminated by the server.

Why can't you let `Faye.Client` resubscribe for you instead of handling this yourself?

**Comment by robinroestenburg** [comment]
> I'm not sure why step 6 is necessary, given the Faye client automatically reinstates its subscriptions for you. 

Indeed, step 6 in the example is performed automatically by the Faye client. It tries to reinstate the subscriptions it had.

> Why can't you let Faye.Client resubscribe for you instead of handling this yourself?

The automatic re-subscription fails, because the signature isn't valid anymore. My Faye server returns `403::Authentication required` in that case. However, the `subscribe` promise has already successfully resolved (because of https://github.com/faye/faye/blob/master/javascript/protocol/client.js#L205); leaving me unable to retrieve a new signature and resubscribe.

That is why I added the hack using `transport:{up,down}`, but I'd rather use Faye's automatic mechanism.

**Comment by jcoglan** [comment]
I'm confused by part of your explanation, which might explain my earlier comment. You say:

> In step 6, the subscription is performed using:
>
> ```js
> client.subscribe('/foo', function(message) {
>   // process message
> }).then(
>   function() { console.log('success'); },
>   function() {
>     console.log('failed');
>     // renew signature and retry
> });

But then you also say:

> Indeed, step 6 in the example is performed automatically by the Faye client.

I don't think the first statement is correct; the re-subscribe that's performed automatically by Faye is done inside [`Client.handshake()`](https://github.com/faye/faye/blob/1.1.1/javascript/protocol/client.js#L103).

The `/meta/subscribe` message generated by that *will* go through your extension code, but it won't re-resolve the promise from your original `subscribe()` call. If the extension is attaching an invalid token, that seems like something to fix in the extension itself.

**Comment by robinroestenburg** [comment]
Ah, yes I see what you mean.

I've created a minimal example from scratch containing the scenarios I was trying to fix: https://github.com/robinroestenburg/faye-invalid-subscription I can see now that I overlooked the fact that all subscription messages pass through the extension. As I had solved scenario 1 through the promise, I was focusing too much on that while trying to get scenario 2 to work.

Thanks for your help on this!
